New project `ouroboros-consensus-cardano-tools`

mgmeier opened the pull request:
A new project `ouroboros-consensus-cardano-tools` is added to the consensus repository. It contains two tools, `db-synthesizer` and `db-analyser`. The former is a new implementation, the latter is a refactoring of the analyser tool from the project `ouroboros-consensus-cardano`.

- `db-synthesizer` synthetically creates a valid ChainDB on-disk; the blocks created do not contain any transactions
- The synthesizer utilizes code ported from the node repository to instantiate `SomeConsensusProtocol` and credentials for block forging
- `db-analyser` is refactored to offer an API for use as a library or as part of a test suite
- Both tools' executables share a common library
- The original analyser remains in place in this PR, so the new executable is currently named `db-analyser2`

A test is provided for the new project where both tools are mutually validated.

deepfire commented on ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal:
```diff
@@ -0,0 +1,186 @@
+  if flag(asserts)
+    ghc-options:       -fno-ignore-asserts
+
+executable db-analyser2
```
I guess we need to negotiate the situation with consensus folks here.

Let's ask them if they agree to drop their analyser and take our modularised version of it.

If they agree -- the drop-and-replacement should be a part of this PR, I'd assume.

cc @nfrisby

deepfire commented on ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal:
```diff
@@ -0,0 +1,186 @@
+                     , ouroboros-consensus-shelley
+                     , ouroboros-network
+
+  other-modules:       Cardano.Api.Key
+                     , Cardano.Api.KeysByron
+                     , Cardano.Api.KeysPraos
+                     , Cardano.Api.KeysShelley
+                     , Cardano.Api.OperationalCertificate
+                     , Cardano.Api.SerialiseUsing
+                     , Cardano.Api.SerialiseTextEnvelope
+                     , Cardano.Node.Protocol.Alonzo
+                     , Cardano.Node.Protocol.Byron
+                     , Cardano.Node.Protocol.Cardano
+                     , Cardano.Node.Protocol.Shelley
```
That's another question of negotiation, I guess.

I reckon that @nfrisby has already implemented the "node config"-based configuration for the analyser -- so maybe we need to join efforts there.

mgmeier replied:
The "node-config" based analyser code is already present in this PR. @nfrisby thankfully notified me when he merged the corresponding feature branch into `master`, so I could port the relevant changes.

A follow-up consideration is part of the draft issue `cardano-tools: code deduplication & future maintenance` in https://github.com/orgs/input-output-hk/projects/22/views/1: using "node-config" in the analyser has opened up room for removing redundant code for both tools.

The decision would be, since that does not change behaviour or API at all, if we want that to be in scope for this PR, or whether it's better suited for a future next step.

deepfire commented on ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs:
```diff
@@ -36,13 +36,16 @@ import           Data.Map.Strict (Map)
 import           Data.Maybe (isJust, mapMaybe)
 import           Data.Proxy
 import qualified Data.Text as Text
+import           Data.Time.Clock (UTCTime)
```
I don't think we still need changes in this file?

mgmeier replied:
Yes indeed, that was an artefact from a previous merge + later rebase. I've removed all those from the PR.

mgmeier commented on ouroboros-consensus-cardano-tools/src/Cardano/Tools/DBSynthesizer/Forging.hs:
```diff
@@ -0,0 +1,219 @@
+
+data ForgeState =
+  ForgeState {
+    currentSlot  :: {-# UNPACK #-} !SlotNo
+  , forged       :: {-# UNPACK #-} !Word64
+  , currentEpoch :: {-# UNPACK #-} !Word64
+  }
+
+initialForgeState :: ForgeState
+initialForgeState = ForgeState 0 0 0
+
+forgingDone :: ForgeLimit -> ForgeState -> Bool
+forgingDone (ForgeLimitSlot s)  = (s == ) . currentSlot
+forgingDone (ForgeLimitBlock b) = (b == ) . forged
+forgingDone (ForgeLimitEpoch e) = (e == ) . currentEpoch
+{-# INLINE forgingDone #-}
```
True... there's however just one single call site for `forgingDone`, and the `ForgeState` is meant to be a "numeric values only" accumulator.
It's not the most CPU intensive code location, and if unpacking/inlining seems overkill here, I can gladly remove it; its impact on onverall runtime on my local machine was there, but minimal...

mgmeier replied:
✅ agreed

dnadales commented on ouroboros-consensus-cardano-tools/src/Cardano/Tools/DBSynthesizer/Orphans.hs:
```diff
@@ -0,0 +1,85 @@
+          , credBulkFile  = f <$> credBulkFile
+          }
+
+-- DUPLICATE: mirroring parsers from cardano-node/src/Cardano/Node/Configuration/POM.hs
```
Silly question maybe, but why can't we use these?

mgmeier replied:
The constraint of this new project was for consensus not to depend on anything in the node repository. So those parts which are relevant for setting up block forging had to be ported to consensus.

dnadales commented on ouroboros-consensus-cardano-tools/src/Cardano/Tools/DBSynthesizer/Run.hs:
```diff
@@ -0,0 +1,163 @@
+                     handleIOExceptT, hoistEither, runExceptT)
+
+import           Control.Tracer (nullTracer)
+{-
```
Dangling comments?

dnadales commented on ouroboros-consensus-cardano-tools/src/Cardano/Tools/DBSynthesizer/Run.hs:
```diff
@@ -0,0 +1,163 @@
+        pure DBSynthesizerConfig {
+              confDbDir     = nfpChainDB
+            , confOptions   = synthOptions
+            , ..
```
I'm trying to vanish wildcards from the code base. I wonder if we want to avoid them here as well.

dnadales commented on ouroboros-consensus-cardano-tools/src/Cardano/Tools/DBSynthesizer/Types.hs:
```diff
@@ -0,0 +1,63 @@
+import           Cardano.Node.Types (ProtocolFilepaths)
+
+
+data NodeConfigStub =
```
We might want to make this alignment consistent with the style in the rest of the code base.

dnadales commented on ouroboros-consensus-cardano-tools/test/Main.hs:
```diff
@@ -0,0 +1,82 @@
+    , limit       = Unlimited
+    }
+
+multiStepTest :: (String -> IO ()) -> Assertion
```
Would it make sense to give this a more descriptive name and add a comment on what the test is doing?

dnadales commented on ouroboros-consensus-cardano-tools/test/Main.hs:
```diff
@@ -0,0 +1,82 @@
+        CardanoBlock b  -> DBAnalyser.analyse testAnalyserConfig b
+        _               -> assertFailure "expexcting test case for Cardano block type"
+
+    resultAnalysis == Just (ResultCountBlock blockCount) @? "wrong number of blocks counted during analysis"
```
Would this report what was the number of synthesized blocks and the number of processed blocks in case of discrepancies? Such information might be useful.

dnadales reviewed (commented):
Thanks for this PR! 

It looks good to me. Here are some things we should address:

- [ ] Decide wether it'd make sense to drop the old db analyser. 
- [ ] Sort import groups according to the consensus style guide.
- [ ] Update the changelog.

mgmeier commented on ouroboros-consensus-cardano-tools/src/Cardano/Tools/DBSynthesizer/Forging.hs:
```diff
@@ -0,0 +1,219 @@
+{-# INLINE forgingDone #-}
+
+
+-- DUPLICATE: runForge mirrors forging loop from ouroboros-consensus/src/Ouroboros/Consensus/NodeKernel.hs
```
There were various reasons why `forkBlockForging` (the top-level definition which contains the forging loop) didn't quite match this requirement:
* It forks a watcher thread, which in turn then kicks off the forging loop nested in there. This approach is not useful here.
* It depends on a mempool / `InternalState` (with peer types), which should not be part of the DB synthesizer.
* It doesn't handle multiple sets of credentials (i.e. `[BlockForging m blk]` from the protocol info), instead relies on forking a thread for each set of credentials. Bulk credenials are very desirable though in this case, for various forgers (based on a staked genesis) to efficiently synthesize a chain.

mgmeier commented on ouroboros-consensus-cardano-tools/app/DBAnalyser/Parsers.hs:
```diff
@@ -0,0 +1,214 @@
+import           Cardano.Tools.DBAnalyser.Block.Shelley
+import           Cardano.Tools.DBAnalyser.Types
+
+import           Data.Foldable (asum)
```
✅ agreed

mgmeier commented on ouroboros-consensus-cardano-tools/app/db-synthesizer.hs:
```diff
@@ -0,0 +1,15 @@
+-- | This tool aims to synthesize a valid ChainDB, replicating cardano-node's UX
```
✅ agreed

mgmeier commented on ouroboros-consensus-cardano-tools/Documentation.md:
```diff
@@ -0,0 +1,110 @@
+* `--show-ebbs` Will print all EBB blocks including their hash, previous block hash and a boolean value whether it is a known EBB (list of known EBBs stored in module `Ouroboros.Consensus.Byron.EBBs`)
+
+* `--store-ledger SLOT_NUMBER` Will store a snapshot of a ledger state under `DB_PATH/ledger/SLOT_NUMBER_db-analyser`. If there is no block under requested slot number, it will create one on the next available slot number (and issue a warning about this fact).
+
```
✅ agreed

mgmeier commented:
@dnadales Thank you for your detailed feedback. I've:

- rebased this branch on current `master`
- removed `db-analyser` from its original location
- adressed all styling and implementation related feedback
- commented on design related feedback
- added documentation

I will remove the draft status of this PR now. Please confirm it's ready for merge, and do mention if any more points should come to mind. Thanks!

dnadales commented on ouroboros-consensus-cardano-tools/src/Cardano/Tools/DBAnalyser/Block/Byron.hs:
```diff
@@ -98,9 +65,9 @@ aBlockOrBoundary fromBoundary fromRegular blk = case blk of
 
 countTxOutputsByron :: Chain.ABlock ByteString -> Int
-countTxOutputsByron Chain.ABlock{..} = countTxPayload bodyTxPayload
+countTxOutputsByron Chain.ABlock{ blockBody } = countTxPayload bodyTxPayload
```
Thanks :heart:

dnadales commented on ouroboros-consensus-cardano-tools/src/Cardano/Tools/DBAnalyser/Block/Cardano.hs:
```diff
@@ -227,7 +241,15 @@ instance Aeson.FromJSON CardanoConfig where
         <> " it must also set it for all previous eras."
 
-    pure $ CardanoConfig{..}
+    pure $ CardanoConfig
```
Minor, we might want to align this. 

Thank you for removing the wildcards!

dnadales reviewed (approved):
Looks good. Thanks! 

Only a minor question: should db-synthesizer be mentioned in the consensus changelog? https://github.com/input-output-hk/ouroboros-network/blob/1caae114f5085ac9fcaf9ef3ca022ececb622695/ouroboros-consensus/docs/interface-CHANGELOG.md

mgmeier commented:
bors r+
